### DISCUSSION TITLE
Collecting `<form>` POST data with Serverless, or with Edge (without Netlify Functions)

### DISCUSSION THREAD

**Discussion by thom4parisot** [11ty/eleventy#2395] [Q&A]
Hello,

I am trying to link a Nextcloud Form with Netlify.
I use Nextcloud API to fetch the form definition (as Global Data), which is then statically rendered. The form has a few fields (first name, last name, event date selector, comment/questions).

With Serverless, I can even populate one field with always up to date options, which is great!

Sadly, I cannot POST the form content directly onto NextCloud, as they expect various CSRF and token elements. So I am aiming at collecting the POST data with Eleventy, and relaying them through Nextcloud HTTP API.

1. I tried collecting the form POST data in a Serverless function, but it is always tracked as a GET call (as if there was an internal redirect somewhere). As I do not use query strings, I have not yet found a way to access the request body.

2. I then tried collecting the form POST data in an Edge function. Sadly, it throws out a "Method Not Allowed" at some point, after the edge function has run.

So far, I have seen the videos and demos showing the use of Serverless and Edge functions with query strings, or by creating a cookie to make _one_ value available in the `serverless` template data.

Has anybody tried something similar with success?
Any hint?

Your help is appreciated,

Thom

**Comment by cfjedimaster** [comment]
Can you share how you POSTed to a serverless function? That is absolutely supported and should work without any issue. In the serverless function, you could do then do your HTTP call to Nextcloud.

**Comment by thom4parisot** [comment]
This is how ends my [`.eleventy.js` config file](https://gitlab.com/solstice.coop/www/-/blob/b1ebaaf0a5dcd4b8afdb43452ff1af820662d6fb/.eleventy.js):

```js
  eleventyConfig.addPlugin(EleventyServerlessBundlerPlugin, {
    name: "form",
    functionsDir: "./netlify/functions/",
    copy: [
      { from: 'css', to: 'css' }
    ]
  })

  return {
    markdownTemplateEngine: "njk",
    htmlTemplateEngine: "njk",
    dir: {
      input: "pages",
      includes: "../_includes",
      layouts: "../_layouts",
      data: "../_data",
    }
  }
```

And this is the output of the [`netlify/functions/form/index.js`](https://gitlab.com/solstice.coop/www/-/blob/b1ebaaf0a5dcd4b8afdb43452ff1af820662d6fb/netlify/functions/form/index.js) when I log the event and the error in the `catch()` section:

```js
{
  event: {
    httpMethod: 'GET',
    path: '/formulaire/',
    rawUrl: 'http://localhost/formulaire/',
    queryStringParameters: {}
  },
  error: Error: No matching URL found for /formulaire/ in {}
      at Serverless.getOutput (/…/node_modules/@11ty/eleventy/src/Serverless.js:174:17)
      at Object.handler (/…/netlify/functions/form/index.js:15:29)
      at BundlerHelper.EleventyServerlessMiddleware (/…/node_modules/@11ty/eleventy/src/Plugins/ServerlessBundlerPlugin.js:223:45)
      at EleventyDevServer.eleventyFolderMiddleware (/…/node_modules/@11ty/eleventy-dev-server/server.js:263:5)
      at EleventyDevServer.onRequestHandler (/…/node_modules/@11ty/eleventy-dev-server/server.js:357:11)
      at Server.emit (node:events:390:28)
      at parserOnIncoming (node:_http_server:951:12)
      at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17) {
    httpStatusCode: 404
  }
}
```

Maybe the error is thrown because the file is located in a subfolder instead of being at the top level folder of the Eleventy project?

And maybe this error is why the POST handling is not done as expected.

The form is deployed here: https://solstice-www-next.netlify.app/formulaire/

**Comment by thom4parisot** [comment]
My bad: I forgot to add the `permalink: { form: '/formulaire/' }` that made the page handled by the serverless middleware.

However, the outcome is similar: request is still shown as a `GET` one, and I still do not have access to the form data without relying on query string parameters :-/

**Reply by cfjedimaster** [reply]
Ok, so you are posting to Eleventy Serverless, not a 'regular' Serverless function like what Netlify has. In that regards, right, you can't POST to it (as far as I know). As you are on Netlify anyway, why not use their Serverless support? I mean technically Eleventy is doing that too, but you get my meaning - a "regular" function, not one tied to Eleventy.

**Reply by thom4parisot** [reply]
Yeah, I might end up doing that. Especially as the final deployment infrastructure will not be Netlify.

Thanks for your help :-)
